server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -186,6 +187,25 @@ public void allocate(RoutingAllocation allocation) {
         balancer.allocateUnassigned();
         balancer.moveShards();
         balancer.balance();
+
+        collectAndRecordNodeWeightStats(balancer, weightFunction, allocation);
```

Comment:
I would like to avoid changing existing allocate function.

BalanceShardAllocator#allocate could be called multiple times per single computation. It looks like we are only interested in the latest or final weights, however this would compute them for each iteration.

I wonder if we could instead introduce a new method (possibly on the interface as well) to compute node weights for the resulting RoutingNodes after executing all iterations? This way the result could be returned immediately rather than be stored on a transient data structure such as Allocation or RoutingNodes.

Reply:
I can do that. we still need to keep track of the last `balancer` and `weightFunction` used for the calculation, simply because we don't know which is the last call to the allocate here. It would save us creating the temporary map.

Reply:
Just an idea after reading it briefly: Is it possible to change the `allocate` method to return a lambda that is essentially `collectAndRecordNodeWeightStats` so that the caller can decide when to call it depending on whether the run has converged?

Reply:
I would suggest not to rely on internal model, but instead compute stats on top of RoutingNodes rather than internal model.

Reply:
That's also where we get the weights. I've pushed https://github.com/elastic/elasticsearch/pull/115854/commits/dad0f67c73d7177a95f2fa08d189e3e3a44d5162, for now.

Reply:
(My preference would be not to recalculate all of this since it is already available where we collect the partial weight from.)